packages/documentsearch/src/searchmodel.ts

```diff
@@ -0,0 +1,308 @@
+        console.error('Failed to update search on document.', reason);
+      });
+    }, searchDebounceTime);
+  }
+
+  /**
+   * Whether the search is case sensitive or not.
+   */
+  get caseSensitive(): boolean {
+    return this._caseSensitive;
+  }
+  set caseSensitive(v: boolean) {
+    if (this._caseSensitive !== v) {
+      this._caseSensitive = v;
+      this.stateChanged.emit();
+      this.refresh();
+    }
+  }
+
+  /**
+   * Current highlighted match index.
+   */
+  get currentIndex(): number | null {
+    return this.searchProvider.currentMatchIndex;
+  }
+
+  /**
+   * A signal emitted when the object is disposed.
+   */
+  get disposed(): ISignal<this, void> {
+    return this._disposed;
+  }
+
+  /**
+   * Filter values.
+   */
+  get filters(): IFilters {
+    return this._filters;
+  }
+  set filters(v: IFilters) {
+    if (!JSONExt.deepEqual(this._filters, v)) {
+      this._filters = v;
+      this.stateChanged.emit();
+      this.refresh();
+    }
+  }
+
+  /**
+   * Filter definitions for the current provider.
+   */
+  get filtersDefinition(): { [n: string]: IFilter } {
+    return this.searchProvider.getFilters?.() ?? {};
+  }
+
+  /**
+   * The initial query string.
+   */
+  get initialQuery(): string {
+    return this._searchExpression || this.searchProvider.getInitialQuery();
+  }
+
+  /**
+   * Whether the document is read-only or not.
+   */
+  get isReadOnly(): boolean {
+    return this.searchProvider.isReadOnly;
+  }
+
+  /**
+   * Parsing regular expression error message.
+   */
+  get parsingError(): string {
+    return this._parsingError;
+  }
+
+  /**
+   * Replacement expression
+   */
+  get replaceText(): string {
+    return this._replaceText;
+  }
+  set replaceText(v: string) {
+    if (this._replaceText !== v) {
+      this._replaceText = v;
+      this.stateChanged.emit();
+    }
+  }
+
+  /**
+   * Search expression
+   */
+  get searchExpression(): string {
+    return this._searchExpression;
+  }
+  set searchExpression(v: string) {
+    if (this._searchExpression !== v) {
+      this._searchExpression = v;
+      this.stateChanged.emit();
+      this.refresh();
+    }
+  }
+
+  /**
+   * Total number of matches.
+   */
+  get totalMatches(): number | null {
+    return this.searchProvider.matchesCount;
+  }
+
+  /**
+   * Whether to use regular expression or not.
+   */
+  get useRegex(): boolean {
+    return this._useRegex;
+  }
+  set useRegex(v: boolean) {
+    if (this._useRegex !== v) {
+      this._useRegex = v;
+      this.stateChanged.emit();
+      this.refresh();
+    }
+  }
+
+  /**
+   * Dispose the model.
+   */
+  dispose(): void {
+    if (this.isDisposed) {
+      return;
+    }
+
+    if (this._searchExpression) {
+      this.endQuery().catch(reason => {
+        console.error(
+          `Failed to end query '${this._searchExpression}.`,
+          reason
+        );
+      });
+    }
+
+    this.searchProvider.stateChanged.disconnect(this.refresh, this);
```

Comment:
Does this cover everything (i.e., disconnection from the state changed signal and also disconnecting all listeners on the disposed signal)?
```suggestion
    Signal.clearData(this);
```